Add quantum machine learning course overview page

frankharkins opened the pull request:
Ready for review, but please do not merge yet.

---
preview url: https://qiskit-org-pr-2362.dcq4xc5i083.us-south.codeengine.appdomain.cloud/textbook-beta/

Fixes https://github.com/Qiskit/qiskit.org/issues/2304  
Fixes https://github.com/Qiskit/qiskit.org/issues/2305

techtolentino commented:
@frankharkins - thanks for this PR.

One quick note: we'll probably want to make the PR _into_ a feature branch. I've setup one for your work:

https://github.com/Qiskit/qiskit.org/tree/feature/qml-overview-page

This way, we're not taking your page, and merging into main (the live site)...
Then, later after we're sure that all the Textbook links are merged/good to go, and after we know your Overview page work has all the right pieces, we'll merge the feature branch, into main.


Please let me know if this makes sense, or just ping us if not.

Thank you!

techtolentino commented on pages/textbook-beta/course/quantum-machine-learning-course.vue:
```diff
@@ -0,0 +1,183 @@
+      </template>
+      <template slot="description">
+        <p
+          v-for="(paragraph, index) in headerDescription"
```
since this `headerDescription` is only 1 paragraph, we don't need to iterate here.

this can be updated to:

`<p v-text="headerDescription" />`

techtolentino commented on pages/textbook-beta/course/quantum-machine-learning-course.vue:
```diff
@@ -0,0 +1,183 @@
+  routeName: string = 'quantum-machine-learning-course'
+
+  headerTitle = 'Quantum machine learning'
+  headerDescription = [`This course contains around eight hours of content, and
```
since the `headerDescription` is only one paragraph, we can remove the array brackets `[` and `]`, leaving us with:

```js
headerDescription = `This course contains around eight hours of content, and
  is aimed at self-learners who are comfortable with undergraduate-level
  mathematics and quantum computing fundamentals. This course will take you
  through key concepts in quantum machine learning, such as parameterized
  quantum circuits, training these circuits, and applying them to basic
  problems. By the end of the course, you'll understand the state of the field,
  and you'll be familiar with recent developments in both supervised and
  unsupervised learning such as quantum kernels and general adversarial networks.
  This course finishes with a project that you can use to showcase what you've
  learnt.`
```

techtolentino commented on constants/appLinks.ts:
```diff
@@ -50,6 +50,7 @@ const TEXTBOOK_BETA_START_LEARNING: GeneralLink = {
 }
 
 const textbookBetaIntroductionCourseStartLearningUrl = 'https://learn.qiskit.org/course/introduction'
+const textbookBetaQuantumMachineLearningCourseStartLearningUrl = 'https://learn.qiskit.org/course/machine-learning'
```
I realize you're following previous naming conventions, but I wonder if it'd be easier to digest if we rename this to something like

`textbookBetaQMLStartLearningUrl`

To be clear, this is not your fault and not a blocker... just my instinct is that these variables are getting a little lengthy, and are now even longer than the string they're representing.

techtolentino commented on pages/textbook-beta/course/quantum-machine-learning-course.vue:
```diff
@@ -0,0 +1,183 @@
+    }
+  ]
+
+  quantumMachineLearningCoursePrerequisites: Prerequisite[] = [
```
Not a blocker for me, since the naming conventions were set before you began this work.. but I'd love to see us as a team, trying to shorten some of these variable names... 

@eddybrando - i'd love to hear your thoughts

frankharkins replied:
For now I've shortened it to `qmlCoursePrerequisites`, hopefully that's a bit better.

techtolentino commented on pages/textbook-beta/course/quantum-machine-learning-course.vue:
```diff
@@ -0,0 +1,183 @@
+      image: 'parameterized-quantum-circuits.png',
+      label: 'Parameterized quantum circuits',
+      segment: { cta: 'parameterized-quantum-circuits', location: 'course' },
+      url: 'https://learn.qiskit.org/course/machine-learning/parameterized-quantum-circuits'
```
Do we know if these `url`s are all finalized?

frankharkins replied:
I can't foresee them changing, only exception is one page that still needs adding. I'll double check everyone's happy with the conventions now though.

techtolentino reviewed (changes requested):
@frankharkins - this is great, thank you.

I had a couple of simple requests, and a few considerations with regards to naming variables... 

🚀

frankharkins commented:
RE links: we decided to go with `machine-learning` over `quantum-machine-learning`, so I've updated the overview page to reflect that.

techtolentino commented on pages/textbook-beta/course/machine-learning-course.vue:
```diff
@@ -0,0 +1,185 @@
+  }
+})
+export default class QuantumMachineLearningCoursePage extends QiskitPage {
+  routeName: string = 'quantum-machine-learning-course'
```
@frankharkins - regarding this update:

> we decided to go with machine-learning over quantum-machine-learning, so I've updated the overview page to reflect that.

I think this should be updated to reflect that.

Also, should everything else remain with `Quantum machine learning`, `qml`, `class="quantum-machine-learning"`, etc?

techtolentino commented on pages/textbook-beta/course/machine-learning-course.vue:
```diff
@@ -0,0 +1,185 @@
+    url: textbookBetaQMLStartLearningUrl,
+    label: 'Start learning',
+    segment: {
+      cta: 'quantum-machine-learning', location: 'course'
```
I feel this should update to `machine-learning`, for consistency.

What do you think?

frankharkins replied:
Yes, thanks. Both were oversights, should be fixed now.

techtolentino commented on pages/textbook-beta/course/machine-learning-course.vue:
```diff
@@ -0,0 +1,185 @@
+    {
+      image: 'introduction.png',
+      label: 'Introduction',
+      segment: { cta: 'qml-introduction', location: 'course' },
```
I'm wondering if we need to be more specific with the `location: course`

such as `location: course-ml` or something.

@eddybrando - do we need more specificity here?

eddybrando replied:
This is OK as is, because `location` in our Segment schema refers to the location in the UI. The "page location" / URL is already available in the default information that gets sent to our analytics.

vabarbosa commented:
@frankharkins @techtolentino 

the preview build for this PR is failing because of a Sass error:

<img width="721" alt="image" src="https://user-images.githubusercontent.com/13156555/148128008-b3d33999-5507-4fd4-a510-d27ccb534674.png">

fyi, i havent tried to investigate the error in depth

vabarbosa commented:
> @frankharkins @techtolentino
> 
> the [preview build](https://github.com/Qiskit/qiskit.org/runs/4706626996?check_suite_focus=true) for this PR is failing because of a Sass error:
> 
> <img alt="image" width="721" src="https://user-images.githubusercontent.com/13156555/148128008-b3d33999-5507-4fd4-a510-d27ccb534674.png">
> 
> fyi, i havent tried to investigate the error in depth

@techtolentino @frankharkins this build errors has been addressed in https://github.com/Qiskit/qiskit.org/pull/2401

how the preview failed again but this time it appears to be a timeout waiting for application to deploy. in any case, the application did deploy and you can preview update here:

https://qiskit-org-pr-2362.dcq4xc5i083.us-south.codeengine.appdomain.cloud/textbook-beta/

JRussellHuffman commented:
Is it possible to add more padding to the right of the image on the home page? Right now, the graphic is right up against the edge of the card.
<img width="546" alt="Screen Shot 2022-01-06 at 12 06 28 PM" src="https://user-images.githubusercontent.com/7633881/148429724-7ef5b35f-d933-44aa-ba60-1aa5979f6127.png">

Also, i think we should take this opportunity to remove the two “under construction” cards altogether. In other words, we should remove the “Traditional Algorithms & Protocols” and “Quantum hardware” section. It sounds like it will probably be late 2022 before those go live, and it’s generally bad practice to have “under construction” up for long periods of time like that.

techtolentino commented:
@JRussellHuffman @frankharkins 

I think the `padding` that Russell mentioned needs to come from the way we export the asset, in my opinion. I'm attaching screenshots showing how different our thumbnail exports are:

![Screen Shot 2022-01-06 at 10 14 54 AM (2)](https://user-images.githubusercontent.com/6276074/148431124-788f7a70-5775-405a-bf59-575301c6bdda.png)

![Screen Shot 2022-01-06 at 10 14 35 AM (2)](https://user-images.githubusercontent.com/6276074/148431134-d30946ef-98e4-48a0-b775-77adb9d2b817.png)

__
Also, we can handle the removal of the `Under construction` cards, in a separate PR, because it might involve removing content, assets, and other code updates.

JRussellHuffman commented:
Good point, the thumbnail export thing should be standardized. I don't normally do the exporting, so maybe this is worth a quick call to talk through? 
My initial thought would be to export it at double the maximum needed size (for retina displays, etc), which is something like 400px by 400px. But there might be more to it than that.

JRussellHuffman commented:
Sounds good removing the under construction tiles in a separate PR.

techtolentino commented:
@frankharkins @JRussellHuffman - added a new issue here, to track the removing of the `Under Construction` cards. 

[That new issue](https://github.com/Qiskit/qiskit.org/issues/2407) _could_ be handled in this same feature branch (`feature/qml-overview-page`), if we want those changes to all roll out together with the qml overview page...

frankharkins commented:
I added a ~20% margin to the thumbnail to match the quantum hardware image, I also added a bit of a margin to the intro course thumbnail while I was there. LMK if that's better.

techtolentino commented:
@frankharkins nicely done 👏🏽

JRussellHuffman commented:
The new margin looks perfect, thanks! @techtolentino It would be best if those changes roll out together.

techtolentino reviewed (approved):
LGTM 👏🏽
